### PR TITLE
Support multiple addon repositories

### PR DESCRIPTION
**Pull request by neoaggelos**
### Summary

Extend the microk8s `status`, `enable` and `disable` commands to support addons from multiple repositories.

Also, introduce the following commands:

```bash
microk8s addons repo list
microk8s addons repo add
microk8s addons repo remove
microk8s addons repo update
```

### Testing

Simple commands to test with a custom addons repository:

```bash
sudo snap install microk8s --classic --channel latest/edge/addon-repos
microk8s addons repo list
microk8s addons repo add neo https://github.com/neoaggelos/microk8s-addons-extras
microk8s addons repo list

microk8s enable neo/nginx
microk8s status
```

### CODE REVIEW

File: scripts/wrappers/disable.py [hunk 1]
```diff
@@ -33,10 +33,7 @@ def disable(addons):
     ensure_started()
     wait_for_ready(timeout=30)
 
-    _, disabled_addons = get_status(get_available_addons(get_current_arch()), True)
-    disabled_addons = {a["name"] for a in disabled_addons}
-
-    xable("disable", addons, disabled_addons)
+    xable("enable", addons)
```
Comment (ktsakalozos) [inline scripts/wrappers/disable.py:36]: Is this supposed to be "enable"?

File: scripts/wrappers/addons.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+        click.echo("Removing {}".format(repo_dir))
+        shutil.rmtree(repo_dir)
+
+    cmd = ["git", "clone", repository, repo_dir]
+    if reference is not None:
+        cmd += ["-b", reference]
+
+    subprocess.check_call(cmd)
+    subprocess.check_call(["chgrp", "microk8s", "-R", repo_dir])
```
Comment (neoaggelos) [inline scripts/wrappers/addons.py:40]: @joedborg Any implications for strict due to this?

File: scripts/wrappers/common/utils.py [hunk 3]
```diff
@@ -187,13 +190,24 @@ def kubectl_get_clusterroles():
...
+            for addon in addons["microk8s-addons"]["addons"]:
+                if arch in addon["supported_architectures"]:
+                    available.append(
+                        {
+                            **addon,
+                            "repository": dir,
+                        }
+                    )
+        except Exception:
+            LOG.exception("could not load addons from %s", addons_yaml)
```
Comment (neoaggelos) [inline scripts/wrappers/common/utils.py:208]: Perhaps printing exceptions here is not the best course of action.

File: scripts/wrappers/common/utils.py [hunk 4]
```diff
@@ -187,13 +190,24 @@ def kubectl_get_clusterroles():
...
+
+    available = sorted(available, key=lambda k: (k["repository"], k["name"]))
```
Comment (neoaggelos) [inline scripts/wrappers/common/utils.py:210]: Sort by repository first, then by name

**Comment by AlexsJones** [comment]
```
sudo snap install microk8s --classic --channel latest/edge/addon-repos
microk8s addons repo list
```

What will the default output?

**Comment by neoaggelos** [comment]
> ```
> sudo snap install microk8s --classic --channel latest/edge/addon-repos
> microk8s addons repo list
> ```
> 
> What will the default output?

Only the core repository, with the `origin` pointing to the snap itself.

```
ubuntu@k8s:~$ microk8s addons repo list
REPO       ADDONS SOURCE
core           31 /snap/microk8s/current/addons/core@684bbe
```

**Comment by AlexsJones** [comment]
We could also symlink `microk8s enable` to `microk8s addons enable` to start to build this muscle memory for this area.

**Comment by balchua** [comment]
This is a great feature.  Now that the addons can refer to a git repo, would there be interest to also point the `core` to a git repo too?

File: scripts/wrappers/addons.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        click.echo("Error: repository '{}' does not exist".format(name), err=True)
+        sys.exit(1)
+
+    click.echo("Removing {}".format(repo_dir))
+    shutil.rmtree(repo_dir)
+
+
+@repository.command("update", help="Update a MicroK8s addons repository")
+@click.argument("name")
+def remove(name: str):
```
Comment (balchua) [inline scripts/wrappers/addons.py:65]: Should this function be named `update` instead of `remove`?

**Review by balchua — COMMENTED** [review]
Is it also assumed that the host system have `git` installed?

**Comment by ktsakalozos** [comment]
When on an lxc container without git installed I got this:
```
root@u4:~# microk8s addons repo add n https://github.com/canonical/microk8s-addons
Cloning into '/var/snap/microk8s/common/addons/n'...
warning: templates not found /usr/share/git-core/templates
fatal: Unable to find remote helper for 'https'
Traceback (most recent call last):
  File "/snap/microk8s/2983/scripts/wrappers/addons.py", line 139, in <module>
    addons(prog_name="microk8s addons")
  File "/snap/microk8s/2983/usr/lib/python3/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/snap/microk8s/2983/usr/lib/python3/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/snap/microk8s/2983/usr/lib/python3/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/snap/microk8s/2983/usr/lib/python3/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/snap/microk8s/2983/usr/lib/python3/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/snap/microk8s/2983/usr/lib/python3/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/snap/microk8s/2983/scripts/wrappers/addons.py", line 39, in add
    subprocess.check_call(cmd)
  File "/snap/microk8s/2983/usr/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['git', 'clone', 'https://github.com/canonical/microk8s-addons', PosixPath('/var/snap/microk8s/common/addons/n')]' returned non-zero exit status 128.
root@u4:~# apt-get install git
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libfreetype6
Use 'apt autoremove' to remove it.
Suggested packages:
  git-daemon-run | git-daemon-sysvinit git-doc git-el git-email git-gui gitk gitweb git-cvs git-mediawiki git-svn
The following NEW packages will be installed:
  git
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 4554 kB of archives.
After this operation, 36.5 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 git amd64 1:2.25.1-1ubuntu3.2 [4554 kB]
Fetched 4554 kB in 2s (2398 kB/s)
Selecting previously unselected package git.
(Reading database ... 30934 files and directories currently installed.)
Preparing to unpack .../git_1%3a2.25.1-1ubuntu3.2_amd64.deb ...
Unpacking git (1:2.25.1-1ubuntu3.2) ...
Setting up git (1:2.25.1-1ubuntu3.2) ...
root@u4:~# microk8s addons repo add n https://github.com/canonical/microk8s-addons
Cloning into '/var/snap/microk8s/common/addons/n'...
remote: Enumerating objects: 469, done.
remote: Counting objects: 100% (469/469), done.
remote: Compressing objects: 100% (329/329), done.
remote: Total 469 (delta 188), reused 405 (delta 137), pack-reused 0
Receiving objects: 100% (469/469), 159.61 KiB | 1.02 MiB/s, done.
Resolving deltas: 100% (188/188), done.
```

**Comment by neoaggelos** [comment]
> root@u4:~# microk8s addons repo add n https://github.com/canonical/microk8s-addons
> Cloning into '/var/snap/microk8s/common/addons/n'...
> warning: templates not found /usr/share/git-core/templates
> fatal: Unable to find remote helper for 'https'

This was due to `git` looking for extra binaries at `/usr/lib/git-core` and template files at `/usr/share/git-core/templates`. The error went away after installing git on the host, because classic confinement allowed the bundled git binary to use the files from the host OS.

This should be fixed now with https://github.com/ubuntu/microk8s/pull/2919/commits/c0f8c9d974316040705b49b23346a653ec33dd0c

**Pull request merged**